### DISCUSSION TITLE
Help with Sprig function syntax in a OMP Template

### DISCUSSION THREAD

**Discussion by mkanet** [JanDeDobbeleer/oh-my-posh#3570] [Troubleshoot] [answered]
Presuming in PowerShell:
```PowerShell
$env:path
```
...will return all the below search paths (delimited with a semicolon character `;`
```
C:\oracle\product\12.1.0\client_1\bin;C:\oracle\product\12.1.0\client_1\bin;C:\WINDOWS\system32;C:\WINDOWS;C:\Program Files\Git\cmd;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;C:\Program Files\Internet Explorer
```

**QUESTION:**
In an OMP Path segment, how do I setup a template to display `\u2606` (⭐character), if the current directory is included in the above `$path` environment variable (separated with semicolon character)?

Don't laugh, below are my best attempts at trying to use `splitList` or `matchP` Sprig functions to achieve this.  However, I am not able to figure out the syntax:

**Attempt 1:**
```
"template": "{{ if .env.Path.SplitList(\";\") .location }}\u2606{{ else }}\uD83D\uDCC2{{ end }} {{ path .Path .Location }}  ",
```

**Attempt 2:**
```
"template": "{{ if contains .env.Path .location }}\u2606{{ else }}\uD83D\uDCC2{{ end }} {{ path .Path .Location }}  ",
```
**Attempt 3:**
```
"template": "{{ if splitList \";\" .env.Path .location }}\u2606{{ else }}\uD83D\uDCC2{{ end }} {{ path .Path .Location }}  "
```

**Attempt 4:**
```
"template": " {{ if matchP .Env.path .location }}\u2606{{ else }}\uD83D\uDCC2{{ end }} {{ path .Path .Location }}  "
```

**NOTE:** If the above is not possible due to limitations.  Could you please suggest the best way to achieve this in OMP?

**Comment by JanDeDobbeleer** [comment]
@JanDeDobbeleer attempt 2 should work, but I think you swapped the variables around. Try:

```
{{ if contains .Location .Env.Path }}
```

**Reply by mkanet** [reply]
@JanDeDobbeleer thank you for the quick reply.  Unfortunately. that still doesn't do it correctly.

Presuming:  $Path = 
```
C:\WINDOWS\system32;C:\Program Files\Internet Explorer
```
![Wrong](https://user-images.githubusercontent.com/811008/224432536-ae626716-bd1e-4f08-968b-8576f9144d85.png)

The problem is that if `.Location = C:\Windows` or `.Location = C:\` they are still contained in `.Env.Path` above.  

I would like to know how to parse/split `.Env.Path` so that it is **only** true if `.Location = C:\Windows\system32` or `.Location = C:\Program Files\Internet Explorer`

I'm not sure how to use [SplitList](https://masterminds.github.io/sprig/string_slice.html) below.  I'm trying to split them via the `;` character; but, can't figure out the correct syntax.  Below, results in **_"invalid template text"_**:
```
{{ if contains .Location .env.Path.SplitList(\";\") }}
```
...or something like that.

**Reply by JanDeDobbeleer** [reply]
@mkanet you need a variable here, I'll create it tomorrow. You need to use splitlist like this BTW:

```
{{ splitList \";\" .Env.Path }}
```

**Reply by mkanet** [reply]
Thanks you!  Looking forward to seeing your solution.

**Reply by mkanet** [reply]
BTW, I tried to create a variable myself.. but, failed miserably LOL.

```
{{ if contains .Location .var.parsedPath }}
```
```
  "version": 2,
  "var": {
    "parsedPath": "{{ splitList \";\" .Env.Path }}"
  }
```

**Reply by JanDeDobbeleer** [reply]
@mkanet here you go. In my case, the `PATH` was split using `:` and casing is very important when referencing env vars.

```json
"template": "{{ $pathList := splitList \":\" .Env.PATH }}{{ if has \"/usr/hello/\" $pathList }}hello {{ end }}"
```

**Reply by mkanet** [reply]
@JanDeDobbeleer Thank you.  It works great.  **There is one small issue.**  If `.Env.PATH` is:

```
C:\WINDOWS:C:\WINDOWS\System32\OpenSSH\:C:\Program Files\Internet Explorer:C:\WINDOWS;C:\Program Files\Git\cmd
```
Your template example will **not** detect `C:\WINDOWS\System32\OpenSSH\` because it ends with a `\` character.  Could you please tell me how to enhance your template example, so it also detects paths ending with `\` character?

**EDIT:**
Maybe we need to use [trimSuffix](https://masterminds.github.io/sprig/strings.html) Spring function somehow?  All my attempts caused _invalid template text_ OMP error.

**Reply by JanDeDobbeleer** [reply]
@mkanet this isn't an easy one as you can't edit what's in the map. I would instead adjust the elements in PATH on Windows to not have a trailing slash.

**Reply by mkanet** [reply]
@JanDeDobbeleer I'm actually designing this prompt for multiple people.  Are you sure there isn't a way to trim the extra `\` from items in `$PathList` below?  Something like this?

```
{{ $pathList := splitList \":\" .Env.Path }}{{ $pathList := trimSuffix \"\\\" $pathList }}
```

**Reply by JanDeDobbeleer** [reply]
@mkanet pathlist is an array, so you'll need to foreach and trim every element before assignment.

**Reply by mkanet** [reply]
@JanDeDobbeleer is there any chance you could please give me an example of how to use 
`foreach` correctly in OMP template?  I can't find any examples to check against.  I can do this in PowerShell, but not sure about how to do it in the OMP template itself.

**Reply by JanDeDobbeleer** [reply]
@mkanet 

```json
"template": "{{ $match := false }}{{ $pathList := splitList \":\" .Env.PATH }}{{ range $val := $pathList }}{{ if eq (trimSuffix \"/\" $val) \"/usr/hello\" }}{{ $match = true }}{{ end }}{{ end }}{{ if $match }}hello {{ end }}"
```

**Reply by mkanet** [reply]
@JanDeDobbeleer please don't get mad.  I understand the logic so far.  However, I'm not sure how to add the trimmed element (and also the other elements) to a new array.  In PowerShell, I would just simply do:  

```PowerShell
foreach ($Element in $PathList) {if ($Element.EndsWith("\")){$NewArray += $Element.TrimEnd("\") } else {$NewArray += $Element}}
```

Once I have a new Array, I can do the rest.

**Reply by JanDeDobbeleer** [reply]
@mkanet you don't need a new array, it wil loop and match a trimmed path. The boolean will tell you if the relevant path was found inside the `PATH` string

**Reply by mkanet** [reply]
I appreciate your patience.  My current code is below, it depends on `$PathList` array to detect which elements are in `.Env.Path`.  I am not sure how to use the code example you gave me in my template below.  The only way I know how to do it is with a newly constructed array that already has its elements trimmed.

```
"template": " {{ $pathList := splitList \";\" .Env.Path }}{{ if has .Location $pathList }}\uD83D\uDD0E{{ else }}\uD83D\uDCC2{{ end }}"
```

**Reply by JanDeDobbeleer** [reply]
@mkanet here you go:

```json
"template": "{{ $match := false }}{{ $pathList := splitList \";\" .Env.PATH }}{{ range $val := $pathList }}{{ if eq (trimSuffix \"/\" $val) .Location }}{{ $match = true }}{{ end }}{{ end }}{{ if $match }}\uD83D\uDD0E{{ else }}\uD83D\uDCC2{{ end }}"
```

**Reply by mkanet** [reply]
@JanDeDobbeleer Unfortunately, the above code produces the error `can't evaluate field Data in type string`.  

**NOTE:**  I can avoid this error by simply changing `.Location` to `\"hello\"`.

**EDIT:**     I tried changing it to below, but it still produces the same error:
```
{{ if eq (trimSuffix \"/\" $val) (.Location | toString) }}
```

**Reply by JanDeDobbeleer** [reply]
@mkanet apparently, the range needs to work with variables only:

```json
"template": "{{ $match := false }}{{ $pwd := .Location }}{{ $pathList := splitList \";\" .Env.PATH }}{{ range $val := $pathList }}{{ if eq (trimSuffix \"/\" $val) $pwd }}{{ $match = true }}{{ end }}{{ end }}{{ if $match }}\uD83D\uDD0E{{ else }}\uD83D\uDCC2{{ end }}"
```

**Reply by mkanet** [reply]
@JanDeDobbeleer the below code doesn't produce an error anymore.  However, $match is never equal to true... not sure why.

```
{{ range $val := $pathList }}{{ if eq (trimSuffix \"/\" $val) $pwd }}{{ $match = true }}
```

I confirmed that `$pwd` is correct.  However, `(trimSuffix \"/\" $val)` doesn't seem to do what we expected, it appears `$val` maybe not the format/type?

**Code:**
```
"{{ $match := false }}{{ $pwd := .Location }}{{ $pathList := splitList \";\" .Env.Path }}{{ range $val := $pathList }}{{ $trimmedpathList := (trimSuffix \"/\" $val) }} {{ $trimmedpathList }}{{ end }}"
```

**Result:**

```
C:\WINDOWS\system32 C:\WINDOWS C:\WINDOWS\System32\WindowsPowerShell\v1.0\ C:\WINDOWS\System32\OpenSSH\ C:\Program Files\Internet Explorer C:\Program Files (x86)\HID Global\ActivClient\
```
Any ideas what might be wrong with `(trimSuffix \"/\" $val)`?  More, specifically, `$val`?

**Reply by JanDeDobbeleer** [reply]
@mkanet this: `{{ $trimmedpathList := (trimSuffix \"/\" $val) }}` doesn't do what you think it does. I just tested this on my mac with a PATH entry that has a trailing `/` and it works as expected. So I'm not sure where it fails on your end.

**Comment by mkanet** [comment]
@JanDeDobbeleer Thank you so much for going out of your way to help.  It's especially hard for me to troubleshoot when it isn't producing any errors.

Do you know how to print `{{ (trimSuffix \"/\" $val) }}` for each element in the below template?  It would tell me why the condition is never `true`; and hopefully tell me what's missing.  Unfortunately I don't have much experience with Sprig functions.

```
"template": "{{ $match := false }}{{ $pwd := .Location }}{{ $pathList := splitList \";\" .Env.PATH }}{{ range $val := $pathList }}{{ if eq (trimSuffix \"/\" $val) $pwd }}{{ $match = true }}{{ end }}{{ end }}{{ if $match }}\uD83D\uDD0E{{ else }}\uD83D\uDCC2{{ end }}" 
```

**EDIT:**  Maybe `$pwd` or `(trimSuffix \"/\" $val)` has `space` characters that need to be trimmed?

**Reply by JanDeDobbeleer** [reply]
@mkanet in my case, trimming this works:

```json
"template": "{{ $match := false }}{{ $pwd := .Location }}{{ $pathList := splitList \";\" .Env.PATH }}{{ range $val := $pathList }}{{ if eq (trimSuffix \"/\" (trim $val)) $pwd }}{{ $match = true }}{{ end }}{{ end }}{{ if $match }}\uD83D\uDD0E{{ else }}\uD83D\uDCC2{{ end }}" 
```

<img width="1130" alt="image" src="https://user-images.githubusercontent.com/2492783/225855342-d208544b-a78d-4885-8029-84181092c412.png">

**Comment by mkanet** [comment]
@JanDeDobbeleer It is hard to believe, but it STILL doesn't work. LOL.  At least we tried our best to guess what's wrong.  At this point, I think it's time to give up trying; especially since there's nothing else left to try.

Anyway, I just want to **thank you for ALL your time to help!**  At least, I learned a little about Sprig Functions. :)

**Reply by JanDeDobbeleer** [reply]
I'll have a look on Windows myself. You never know there's something obvious I'm missing.

**Reply by mkanet** [reply]
Wow.  Really?  I would really appreciate it! I have to admit, it is really bugging me that I couldn't figure it out.  Hopefully, you'll have better luck.

**Reply by JanDeDobbeleer** [reply, marked as answer]
@mkanet this works for me on Windows:

```json
"template": "{{ $match := false }}{{ $pwd := .Location }}{{ $pathList := splitList \";\" .Env.Path }}{{ range $val := $pathList }}{{ if eq (trimSuffix \"\\\\\" (trim $val)) $pwd }}{{ $match = true }}{{ end }}{{ end }}{{ if $match }}\uD83D\uDD0E{{ else }}\uD83D\uDCC2{{ end }}"
```

![image](https://user-images.githubusercontent.com/2492783/226098911-b7f11ff7-630a-4010-b987-77465516bf67.png)

**Reply by mkanet** [reply]
@JanDeDobbeleer I'm so sorry for replying so late.  I misunderstood your last comment.  I didn't realize the code you posted was useable in my environment too.  I thought it would only work on your Windows PC.

Anyway, it worked PERFECTLY!    I went a step further and added the 🏠 icon too if in the `.Env.USERPROFILE` path.  It turned my template into _probably_ the longest oh-my-posh template!!😮:

```
"template": "  {{ $match := false }}{{ $pwd := .Location }}{{ $pathList := splitList \";\" .Env.Path }}{{ range $val := $pathList }}{{ if eq (trimSuffix \"\\\\\" (trim $val)) $pwd }}{{ $match = true }}{{ end }}{{ end }}{{ if $match }}\uD83D\uDD0E{{ else if eq .Env.USERPROFILE .Location}}\uD83C\uDFE0{{ else }}\uD83D\uDCC2{{ end }} {{ path .Location .Location }} ",
```

Anyway, thank you so much for spending so much time helping me with this!  I really appreciate it!
